src/persist/src/s3.rs

```diff
@@ -357,179 +370,127 @@ impl Blob for S3Blob {
             Err(SdkError::ServiceError(err)) if err.err().is_no_such_key() => return Ok(None),
             Err(err) => return Err(ExternalError::from(anyhow!("s3 get meta err: {}", err))),
         };
-        let num_parts = if first_part.parts_count() == 0 {
-            // For a non-multipart upload, parts_count will be 0. The rest of
-            // the code works perfectly well if we just pretend this was a
-            // multipart upload of 1 part.
-            1
-        } else {
-            first_part.parts_count()
+
+        // Get the remaining number of parts
+        let num_parts = match first_part.parts_count() {
+            // For a non-multipart upload, parts_count will be 0. The rest of  the code works
+            // perfectly well if we just pretend this was a multipart upload of 1 part.
+            0 => 1,
+            // For any positive value greater than 0, just return it.
+            parts @ 1.. => parts,
+            // A negative value is invalid.
+            bad => {
+                assert!(bad < 0);
+                return Err(anyhow!("unexpected number of s3 object parts: {}", bad).into());
+            }
         };
+
         trace!(
-            "s3 download first header took {:?} ({} parts)",
+            "s3 download first header took {:?} ({num_parts} parts)",
             start_overall.elapsed(),
-            num_parts
         );
-        if num_parts < 1 {
-            return Err(anyhow!("unexpected number of s3 object parts: {}", num_parts).into());
-        }
 
+        // Get handle to our runtime so we can concurrently fetch the remaining parts.
         let async_runtime = AsyncHandle::try_current().map_err(anyhow::Error::msg)?;
 
-        // Continue to fetch the first part's body while we fetch the other
-        // parts.
-        let start_part_body = Instant::now();
-        let mut part_bodies = Vec::new();
-        let first_body_len = usize::try_from(first_part.content_length)
-            .map_err(|err| anyhow!("unexpected s3 content_length: {}", err))?;
-        let mut total_body_len = first_body_len;
-        part_bodies.push(
-            // TODO: Add the key and part number once this can be annotated with
-            // metadata.
-            async_runtime.spawn_named(
-                || "persist_s3blob_get_body",
-                first_part
-                    .body
-                    .collect()
-                    .map(move |res| (start_part_body.elapsed(), first_body_len, res)),
-            ),
-        );
+        // Fetch the first part's body while we fetch the other parts.
+        let min_body_elapsed_ = Arc::clone(&min_body_elapsed);
+        let first_body_fut = async move {
+            let body_start = Instant::now();
+            let result = first_part
+                .body
+                .collect()
+                .map_err(|err| Error::from(format!("s3 get first body err: {}", err)));
+            let elapsed = body_start.elapsed();
```

Comment:
is this the same? I think you need to await the `collect()` future before calling `elapsed()`, no?